Filter nethealth data

bernardjkim opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
This PR updates the nethealth checker to filter incoming network data. The nethealth app may persist metrics for peers that are no longer members of the cluster. Metrics for these peers need to be filtered and removed.

## Implementation
The nethealth-checker now filters incoming nethealth data through `filterNetData`. This function removes nethealth data for peers that are no longer members of the clutser. Cluster membership is determined by querying k8s nodes.

Also updated `getNethealthAddr` function to query pods by `NodeName` instead of `HostIP`.

Previous implementation https://github.com/gravitational/satellite/pull/191. 

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Write tests
- [x] Perform manual testing
- [x] Address review feedback

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->
**Before changes**

- Setup multi-node cluster.
- Degrade nethealth checker using iptables to drop traffic from one node to another
```
[vagrant@node-1 ~]$ sudo iptables -A INPUT -p udp -s 172.28.128.102 -j DROP
```

- Once nodes are in a degraded state due to nethealth checker remove one of the degraded nodes.
```
[vagrant@node-1 ~]$ sudo gravity remove node-2 --force
```

- After the node is removed, the nethealth check is stuck in this state.
```
[vagrant@node-1 ~]$ sudo gravity status
Cluster name:           dev.test
Cluster status:         degraded
[...]
Cluster nodes:
    Masters:
        * node-1 (172.28.128.101, node)
            Status:     degraded
            [×]         overlay packet loss for node 172.28.128.102 is higher than the allowed threshold of 20.00%: 100.00% ()
        * node-3 (172.28.128.103, node)
            Status:     healthy
```

**After changes** 
Using gravity 7.0.4 with dev planet build containing changes from this satellite branch.

- Setup mutli-node cluster and block udp between `node-1` and `node-2`. 
```
[vagrant@node-1 ~]$ sudo gravity status
Cluster name:           dev.test
Cluster status:         degraded
Cluster image:          telekube, version 7.0.4
Gravity version:        7.0.4 (client) / 7.0.4 (server)
[...]
Cluster nodes:
    Masters:
        * node-3 / 172.28.128.103 / node
            Status:     healthy
        * node-1 / 172.28.128.101 / node
            Status:     degraded
            [×]         overlay packet loss for node 172.28.128.102 is higher than the allowed threshold of 20.00%: 100.00%
        * node-2 / 172.28.128.102 / node
            Status:     degraded
            [×]         overlay packet loss for node 172.28.128.101 is higher than the allowed threshold of 20.00%: 100.00%
[ERROR]: degraded
```

- After we remove `node-2` from the cluster `node-1` should no longer be degraded.
```
[vagrant@node-1 ~]$ sudo gravity status
Cluster name:           dev.test
Cluster status:         active
Cluster image:          telekube, version 7.0.4
Gravity version:        7.0.4 (client) / 7.0.4 (server)
[...]
Cluster nodes:
    Masters:
        * node-3 / 172.28.128.103 / node
            Status:     healthy
        * node-1 / 172.28.128.101 / node
            Status:     healthy
```

Nethealth check functions similarly when running on worker node.

- Rejoin `node-2` with role `knode` and block udp between `node-2` and `node-3`.
```
[vagrant@node-2 ~]$ sudo gravity status
Cluster name:           dev.test
Cluster status:         degraded
Cluster image:          telekube, version 7.0.4
Gravity version:        7.0.4 (client) / 7.0.4 (server)
[...]
Cluster nodes:
    Masters:
        * node-3 / 172.28.128.103 / node
            Status:     degraded
            [×]         overlay packet loss for node 172.28.128.102 is higher than the allowed threshold of 20.00%: 100.00%
        * node-1 / 172.28.128.101 / node
            Status:     healthy
    Nodes:
        * node-2 / 172.28.128.102 / knode
            Status:     degraded
            [×]         overlay packet loss for node 172.28.128.103 is higher than the allowed threshold of 20.00%: 100.00%
[ERROR]: degraded
```

- After removing `node-3` `node-2` eventually returns to a healthy state.
```
[vagrant@node-1 ~]$ sudo gravity status
Cluster name:           dev.test
Cluster status:         active
Cluster image:          telekube, version 7.0.4
Gravity version:        7.0.4 (client) / 7.0.4 (server)
[...]
Cluster nodes:
    Masters:
        * node-1 / 172.28.128.101 / node
            Status:     healthy
    Nodes:
        * node-2 / 172.28.128.102 / knode
            Status:     healthy
```

bernardjkim committed:
Select pods by node name instead of host ip

bernardjkim committed:
Filter by k8s nodes

bernardjkim committed:
Add filter test cases

r0mant commented on monitoring/nethealth.go:
```diff
@@ -153,22 +161,24 @@ func (c *nethealthChecker) check(ctx context.Context, reporter health.Reporter)
+	}
+
+	pod := pods.Items[0]
+	if pod.Status.PodIP == "" {
```
Btw, I wonder if we should also look at the pod status? I.e. only consider Running pods. Otherwise the pod for example may still be starting up and not able to serve requests.

r0mant commented on monitoring/nethealth.go:
```diff
@@ -396,6 +406,34 @@ func getPeerName(labels []*dto.LabelPair) (peer string, err error) {
 	return "", trace.NotFound("unable to find %s label", peerLabel)
 }
 
+// filterNetData filters the networkData. Nethealth may retain metrics for nodes
```
Nit: The parameter is called "netData".

r0mant commented on monitoring/nethealth.go:
```diff
@@ -396,6 +406,34 @@ func getPeerName(labels []*dto.LabelPair) (peer string, err error) {
+	for _, node := range nodes {
+		data, exists := netData[node.Name]
+		if !exists {
+			log.WithField("node", node.Name).Warn("missing nethealth data for peer")
```
Nit: Start log message with capital letter and end with a dot.

r0mant commented on monitoring/nethealth_test.go:
```diff
@@ -296,6 +350,15 @@ func (s *NethealthSuite) newPacketLoss(values ...float64) []float64 {
 	return packetLoss
 }
 
+// newTestNode constructs a new ndoe with the provided node name.
```
node

r0mant reviewed (commented):
Btw, I've just noticed in this error message we give:

```
overlay packet loss for node 172.28.128.102 is higher than the allowed threshold of 20.00%: 100.00%
```

Can we get rid of `.00` everywhere and just say `20%`, `100%`, etc.? :)

bernardjkim committed:
Address PR feedback

- No longer displays decimal point
- Verify nethealth pod is running
- Fix godoc and logs

bernardjkim committed:
Change nethealth check to a warning
